alea/utils.py

```diff
@@ -215,3 +327,222 @@ def clip_limits(value) -> Tuple[float, float]:
+    argument into strings.
+
+    Args:
+        value_type: type of the value, might be float, int, etc.
+        target_type: type of the target, might be Optinal, Union, etc.
+
+    """
+    if get_origin(target_type) is Union:
+        # If the target type is a Union (like Optional)
+        return any(can_assign_to_typing(value_type, t) for t in get_args(target_type))
+    else:
+        if get_origin(target_type):
+            return issubclass(value_type, get_origin(target_type))
+        else:
+            return issubclass(value_type, target_type)
+
+
+def add_i_batch(filename: str) -> str:
+    """Add i_batch to filename."""
+    if "i_batch" in filename:
+        raise ValueError("i_batch already in filename")
+    fpat_split = os.path.splitext(filename)
+    return fpat_split[0] + "_{i_batch:d}" + fpat_split[1]
+
+
+def can_expand_grid(variations: dict) -> bool:
+    """Check if variations can be expanded into a grid.
+
+    Example:
+        >>> can_expand_grid({'a': [1, 2], 'b': [3, 4]})
+        True
+
+    """
+
+    # check if all values are lists or no values is list
+    is_list = [isinstance(value, list) for value in variations.values()]
+    if {True, False}.issubset(is_list):
+        raise ValueError(
+            "If some values in variations are lists, "
+            "all values must be lists or no values is list. "
+            f"But you are mixing lists and non-lists in {variations}."
+        )
+    if all(is_list):
+        return True
+    else:
+        return False
+
+
+def expand_grid_dict(variations: List[Union[dict, str]]) -> List[Union[dict, str]]:
+    """Expand dict into a list of dict, according to the itertools.product method, if necessary.
+
+    Args:
+        variations (list): variations to be expanded
+
+    Example:
+        >>> expand_grid_dict(["free", {"a": 1, "b": 3}, {"a": [1, 2], "b": [3, 4]}])
+        [
+            "free",
+            {"a": 1, "b": 3},
+            {"a": 1, "b": 3},
+            {"a": 1, "b": 4},
+            {"a": 2, "b": 3},
+            {"a": 2, "b": 4},
+        ]
+
+    """
+
+    result = cast(List[Union[dict, str]], [])
+    for v in variations:
+        # convert str to list first
+        if isinstance(v, dict):
+            v = evaluate_numpy_scipy_expression_in_dict(v)
+        # expand to grid if necessary
+        if isinstance(v, dict) and can_expand_grid(v):
+            result += convert_to_vary(v)
+        else:
+            result.append(v)
+    return result
+
+
+def convert_variations(variations: dict, iteration) -> list:
+    """Convert variations to a list of dict, according to the iteration method.
+
+    Args:
+        variations (dict): variations to be converted
+        iteration: iteration method, either zip or itertools.product
+
+    Returns:
+        list: a list of dict
+
+    """
+
+    # evaluate numpy and scipy expression in variations
+    variations = evaluate_numpy_scipy_expression_in_dict(variations)
```

Comment:
Cab you give an example when this would be used?

Reply:
Like here https://github.com/XENONnT/alea/blob/d280821cdc476615b1d3ff386eb52063a533a82d/alea/utils.py#L389

```
expand_grid_dict(["free", {"a": 1, "b": 3}, {"a": np.linspace(0, 1, 11), "b": [3, 4]}]) 
```

Reply:
Maybe could you just add a similar example to the docstring of this method? I think it would be super useful for the future 😅

Reply:
Added.